source/common/http/filter_manager.cc

```diff
@@ -791,12 +796,36 @@ void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMa
       (*entry)->getSavedRequestMetadata()->emplace_back(std::move(metadata_map_ptr));
       return;
     }
-
+    state_.filter_call_state_ |= FilterCallState::DecodeMetadata;
     FilterMetadataStatus status = (*entry)->handle_->decodeMetadata(metadata_map);
-    ASSERT(!hasPreparedLocalReply(), "Sending Local Reply from Metadata is not yet supported.");
+    state_.filter_call_state_ &= ~FilterCallState::DecodeMetadata;
+
     ENVOY_STREAM_LOG(trace, "decode metadata called: filter={} status={}, metadata: {}", *this,
                      (*entry)->filter_context_.config_name, static_cast<uint64_t>(status),
                      metadata_map);
+    if (state_.decoder_filter_chain_aborted_) {
```

Comment:
Ah, so one of the reasons I asked you to e2e test with an abortIFSeesHeader filter is I don't think this works for the upstream filter chain.  decoder_filter_chain_aborted is fundamentally a downstream boolean as it's triggered on local reply.  local replies only hit the downstream filter chain, so I think what will happen if you execute this path on the upstream filter chain is we'll send a local reply but keep iterating on metadata.

Having looked at the filter you used it doesn't crash on metadata.  Can you update that and see if my concern is valid?  If so I'd love another rename TODO for either you or me to rename these booleans  to be more clear 
/wait

Reply:
oh hey look I already safeguarded against this mistake and forgot about it.
https://github.com/envoyproxy/envoy/blob/main/source/common/router/upstream_request.cc#L69